src/fs/file.rs

```diff
@@ -170,6 +170,89 @@ impl File {
         op.read().await
     }
 
+    /// Read the exact number of bytes required to fill `buf` at the specified
+    /// offset from the file.
+    ///
+    /// This function reads as many as bytes as necessary to completely fill the
+    /// specified buffer `buf`.
+    ///
+    /// # Return
+    ///
+    /// The method returns the operation result and the same buffer value passed
+    /// as an argument.
+    ///
+    /// If the method returns [`Ok(())`], then the read was successful.
+    ///
+    /// # Errors
+    ///
+    /// If this function encounters an error of the kind [`ErrorKind::Interrupted`]
+    /// then the error is ignored and the operation will continue.
+    ///
+    /// If this function encounters an "end of file" before completely filling
+    /// the buffer, it returns an error of the kind [`ErrorKind::UnexpectedEof`].
+    /// The buffer is returned on error.
+    ///
+    /// If this function encounters any form of I/O or other error, an error
+    /// variant will be returned. The buffer is returned on error.
+    ///
+    /// # Examples
+    ///
+    /// ```no_run
+    /// use tokio_uring::fs::File;
+    ///
+    /// fn main() -> Result<(), Box<dyn std::error::Error>> {
+    ///     tokio_uring::start(async {
+    ///         let f = File::open("foo.txt").await?;
+    ///         let buffer = vec![0; 10];
+    ///
+    ///         // Read up to 10 bytes
+    ///         let (res, buffer) = f.read_exact_at(buffer, 0).await;
+    ///         res?;
+    ///
+    ///         println!("The bytes: {:?}", buffer);
+    ///
+    ///         // Close the file
+    ///         f.close().await?;
+    ///         Ok(())
+    ///     })
+    /// }
+    /// ```
+    ///
+    /// [`ErrorKind::Interrupted`]: std::io::ErrorKind::Interrupted
+    /// [`ErrorKind::UnexpectedEof`]: std::io::ErrorKind::UnexpectedEof
+    pub async fn read_exact_at<T: IoBufMut>(
+        &self,
+        mut buf: T,
+        pos: u64,
+    ) -> crate::BufResult<(), T> {
+        let buf_len = buf.bytes_init();
```

Comment:
I think we want to use `bytes_total()` here to read in uninitialized memory? Though, this would change the API. E.g. passing in `Vec::with_capacity(100)` would read 100 bytes.

The `slice(..)` API could be used for limiting the length to read.

Reply:
I did recognize that the `bytes_total()` instead of `bytes_init()` is used in the read operation in `tokio-uring`. But the `std::io::Read::read_exact`'s parameter is a `&mut [u8]`, whose length is immutable.
Maybe it's time to choose  consistency (in this crate) or conveyance  (between std and other libraries).
I think it's good to keep consistency in this crate and inform users the differences.

Reply:
@carllerche Does the current form of the code satisfy your concern from Aug 2021?

Reply:
@Noah-Kennedy Looking at the latest version of this code, it does use bytes_total() for the read and only the bytes_init() for the write. So I think the concern was taken to heart. Want to have this PR ignore a further review by @carllerche? Then it can be merged.